src/Ubidots.cpp

```diff
@@ -0,0 +1,227 @@
+                      dot_timestamp_millis);
+}
+
+/**
+ * Sends data to Ubidots
+ * @arg device_label [Mandatory] device label where the dot will be stored
+ * @arg device_name [optional] Name of the device to be created (supported only
+ * for TCP/UDP)
+ * @arg flags [Optional] Particle publish flags for webhooks
+ */
+bool Ubidots::send() { return send(_defaultDeviceLabel, _defaultDeviceLabel); }
+
+bool Ubidots::send(const char *device_label) {
+  return send(device_label, device_label);
+}
+
+bool Ubidots::send(const char *device_label, const char *device_name) {
+  return _cloudProtocol->send(device_label, device_name);
+}
+
+float Ubidots::get(const char *device_label, const char *variable_label) {
+  if (_debug) {
+    Serial.println("Constructing get Request!");
+  }
+
+  _cloudProtocol->get(device_label, variable_label);
+}
+
+void Ubidots::setDebug(bool debug) {
+  _debug = debug;
+  _cloudProtocol->setDebug(debug);
+}
+
+/*
+ * Adds to the context structure values to retrieve later it easily by the user
+ */
+
+void Ubidots::addContext(char *key_label, char *key_value) {
+  (_context + _current_context)->key_label = key_label;
+  (_context + _current_context)->key_value = key_value;
+  _current_context++;
+  if (_current_context >= MAX_VALUES) {
+    if (_debug) {
+      Serial.println(F("You are adding more than the maximum of consecutive "
+                       "key-values pairs"));
+    }
+    _current_context = MAX_VALUES;
+  }
+}
+
+/*
+ * Retrieves the actual stored context properly formatted
+ */
+
+void Ubidots::getContext(char *context_result) {
+  getContext(context_result, _iotProtocol);
+}
+
+void Ubidots::getContext(char *context_result, IotProtocol iotProtocol) {
+  // TCP context type
+  if (iotProtocol == UBI_TCP || iotProtocol == UBI_UDP) {
+    sprintf(context_result, "");
+    for (uint8_t i = 0; i < _current_context;) {
+      sprintf(context_result, "%s%s=%s", context_result,
+              (_context + i)->key_label, (_context + i)->key_value);
+      i++;
+      if (i < _current_context) {
+        sprintf(context_result, "%s$", context_result);
+      } else {
+        sprintf(context_result, "%s", context_result);
+        _current_context = 0;
+      }
+    }
+  }
+
+  // HTTP context type
+  if (iotProtocol == UBI_HTTP) {
+    sprintf(context_result, "");
+    for (uint8_t i = 0; i < _current_context;) {
+      sprintf(context_result, "%s\"%s\":\"%s\"", context_result,
+              (_context + i)->key_label, (_context + i)->key_value);
+      i++;
+      if (i < _current_context) {
+        sprintf(context_result, "%s,", context_result);
+      } else {
+        sprintf(context_result, "%s", context_result);
+        _current_context = 0;
+      }
+    }
+  }
+}
+
+bool Ubidots::serverConnected() { return _cloudProtocol->serverConnected(); }
+
+void Ubidots::getDeviceIMEI(char IMEI[]) {
+  const uint8_t IMEI_SIZE = 20;
+
+  GPRS *gprs = GPRS::getInstance();
```

Comment:
why do we need another client? we should use the static client instance

Reply:
The GPRS Class implements the singleton pattern so, wherever you call for getInstance() it will return the pointer from the object wherever was created. So, I'm using the same object from UbiTCP.

Reply:
There is no way to pass the GPRS object from UbiTCP to this class without make it static.